include/zephyr/drivers/timer/system_timer.h

```diff
@@ -109,19 +109,14 @@ extern void sys_clock_announce(int32_t ticks);
  */
 extern uint32_t sys_clock_elapsed(void);
 
-#if defined(CONFIG_SYS_CLOCK_EXISTS) && \
-	defined(CONFIG_SYSTEM_TIMER_HAS_DISABLE_SUPPORT) || \
-	defined(__DOXYGEN__)
 /**
  * @brief Disable system timer.
  *
- * This function is a no-op if the system timer does not have the capability
- * of being disabled.
+ * @note Not all system timer drivers has the capability of being disabled.
+ * The config @kconfig{CONFIG_SYSTEM_TIMER_HAS_DISABLE_SUPPORT} can be used to
+ * check if the system timer has the capability of being disabled.
  */
 extern void sys_clock_disable(void);
-#else
-static inline void sys_clock_disable(void) {}
-#endif /* CONFIG_SYSTEM_TIMER_HAS_DISABLE_SUPPORT */
```

Comment:
Not that there's anything wrong with the code, but AFAIK the spirit of that coding guideline rule is that we shouldn't have an API whose expression depends on kconfig.  Something that is legal Zephyr code should be legal in all configurations.

What we had here was a situation where the **definition** of a function was being selected (as a noop) based on kconfig.  That's just a straightforward optimization.  The function was declared compatibly in all cases.

That said, if there was only one place the function was being called, it makes much more sense to put the kconfig check there.